### PR TITLE
r/glacier_vault: fix policy diff + read after create + refactor tests

### LINKED ISSUE #12619: Glacier vault import fails
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

0.12.24

### Affected Resource(s)

* aws_glacier_vault

### Terraform Configuration Files

```hcl
resource "aws_glacier_vault" "photos" {
  name = "photos"
}
```

### Debug Output

https://gist.githubusercontent.com/artamonovkirill/528f00dd39c908586ed4e68223e2f7cc/raw/477a01e35563e5a27d0cc0c805f0cc430bde73a0/debug.out

### Panic Output

https://gist.githubusercontent.com/artamonovkirill/528f00dd39c908586ed4e68223e2f7cc/raw/477a01e35563e5a27d0cc0c805f0cc430bde73a0/crash.log

### Expected Behavior

Glacier vault is imported.

### Actual Behavior

Terraform crashed.

### Steps to Reproduce

`terraform import aws_glacier_vault.photos photos`

### Important Factoids

No.

### References

n/a

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #12619 (possibly)
Relates #13826

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_glacier_vault: remove from state when vault does not exist.
resource_aws_glacier_vault: add plan time validation to `sns_topic` in the `notification` block.
resource_aws_glacier_vault: add plan time validation to `events` in the `notification` block.
resource_aws_glacier_vault: fix diff for `access_policy`.
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSGlacierVault_'
--- PASS: TestAccAWSGlacierVault_basic (51.11s)
--- PASS: TestAccAWSGlacierVault_tags (145.51s)
--- PASS: TestAccAWSGlacierVault_disappears (32.80s)
--- PASS: TestAccAWSGlacierVault_notification (134.42s)
--- PASS: TestAccAWSGlacierVault_policy (79.08s)
```

### CODE REVIEW

File: aws/resource_aws_glacier_vault.go [hunk 1]
```diff
@@ -165,14 +160,14 @@ func resourceAwsGlacierVaultRead(d *schema.ResourceData, meta interface{}) error
 	}
 
 	log.Printf("[DEBUG] Getting the access_policy for Vault %s", d.Id())
-	pol, err := glacierconn.GetVaultAccessPolicy(&glacier.GetVaultAccessPolicyInput{
+	pol, err := conn.GetVaultAccessPolicy(&glacier.GetVaultAccessPolicyInput{
 		VaultName: aws.String(d.Id()),
 	})
 
-	if awserr, ok := err.(awserr.Error); ok && awserr.Code() == "ResourceNotFoundException" {
```
Comment (ewbankkit) [inline aws/resource_aws_glacier_vault.go:172]: Maybe?

```go
	if isAWSErr(err, glacier.ErrCodeResourceNotFoundException, "") {
		d.Set("access_policy", "")
	} else if err != nil 
		return fmt.Errorf("error getting access policy for Glacier Vault (%s): %s", d.Id(), err)
	} else if pol != nil && pol.Policy != nil {
		policy, err := structure.NormalizeJsonString(aws.StringValue(pol.Policy.Policy))
		if err != nil {
			return fmt.Errorf("access policy contains an invalid JSON: %s", err)
		}
		d.Set("access_policy", policy)
	}
```

Comment (DrFaust92) [reply aws/resource_aws_glacier_vault.go:172]: added

**Comment by DrFaust92** [comment]
rebased + retested

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSGlacierVault_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSGlacierVault_ -timeout 120m
=== RUN   TestAccAWSGlacierVault_basic
=== PAUSE TestAccAWSGlacierVault_basic
=== RUN   TestAccAWSGlacierVault_notification
=== PAUSE TestAccAWSGlacierVault_notification
=== RUN   TestAccAWSGlacierVault_policy
=== PAUSE TestAccAWSGlacierVault_policy
=== RUN   TestAccAWSGlacierVault_tags
=== PAUSE TestAccAWSGlacierVault_tags
=== RUN   TestAccAWSGlacierVault_disappears
=== PAUSE TestAccAWSGlacierVault_disappears
=== CONT  TestAccAWSGlacierVault_basic
=== CONT  TestAccAWSGlacierVault_tags
=== CONT  TestAccAWSGlacierVault_disappears
=== CONT  TestAccAWSGlacierVault_policy
=== CONT  TestAccAWSGlacierVault_notification
--- PASS: TestAccAWSGlacierVault_disappears (19.35s)
--- PASS: TestAccAWSGlacierVault_basic (28.48s)
--- PASS: TestAccAWSGlacierVault_policy (31.39s)
--- PASS: TestAccAWSGlacierVault_tags (66.01s)
--- PASS: TestAccAWSGlacierVault_notification (71.82s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	71.875s
```

**Comment by DrFaust92** [comment]
Refactored:

```
--- PASS: TestAccAWSGlacierVault_notification (152.71s)
--- PASS: TestAccAWSGlacierVault_policy (135.11s)
--- PASS: TestAccAWSGlacierVault_tags (133.06s)
--- PASS: TestAccAWSGlacierVault_disappears (41.30s)
```

File: aws/resource_aws_glacier_vault_test.go [hunk 2]
```diff
@@ -84,9 +85,14 @@ func TestAccAWSGlacierVault_basic(t *testing.T) {
 		CheckDestroy: testAccCheckGlacierVaultDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccGlacierVault_basic(rInt),
+				Config: testAccGlacierVaultBasicConfig(rName),
 				Check: resource.ComposeTestCheckFunc(
-					testAccCheckGlacierVaultExists(resourceName),
+					testAccCheckGlacierVaultExists(resourceName, &vault),
+					resource.TestCheckResourceAttr(resourceName, "tags.#", "0"),
```
Comment (bflad) [inline aws/resource_aws_glacier_vault_test.go:91]:
```suggestion
					resource.TestCheckResourceAttr(resourceName, "tags.%", "0"),
```

**Review by bflad — APPROVED** [review]
Looks good to me, thank you, @DrFaust92 🚀 

Output from acceptance testing:

```
--- PASS: TestAccAWSGlacierVault_disappears (12.48s)
--- PASS: TestAccAWSGlacierVault_basic (17.66s)
--- PASS: TestAccAWSGlacierVault_policy (41.22s)
--- PASS: TestAccAWSGlacierVault_tags (43.39s)
--- PASS: TestAccAWSGlacierVault_notification (46.12s)
```

File: aws/resource_aws_glacier_vault_test.go [hunk 3]
```diff
@@ -239,47 +348,119 @@ func testAccCheckGlacierVaultDestroy(s *terraform.State) error {
...
+}
+`, rName)
+}
 
-  tags = {
-    Test = "Test1"
-  }
+func testAccGlacierVaultPolicyConfig(rName string) string {
+	return fmt.Sprintf(`
+data "aws_region" "current" {}
```
Comment (bflad) [inline aws/resource_aws_glacier_vault_test.go:378]:
```suggestion
data "aws_partition" "current" {}

data "aws_region" "current" {}
```

File: aws/resource_aws_glacier_vault_test.go [hunk 4]
```diff
@@ -239,47 +348,119 @@ func testAccCheckGlacierVaultDestroy(s *terraform.State) error {
...
+          "Principal": {
+             "AWS": ["*"]
+          },
+          "Effect":"Allow",
+          "Action": [
+             "glacier:InitiateMultipartUpload",
+             "glacier:AbortMultipartUpload",
+             "glacier:CompleteMultipartUpload"
+          ],
+          "Resource": ["arn:aws:glacier:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:vaults/%[1]s"]
```
Comment (bflad) [inline aws/resource_aws_glacier_vault_test.go:400]:
```suggestion
          "Resource": ["arn:${data.aws_partition.current.partition}:glacier:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:vaults/%[1]s"]
```

File: aws/resource_aws_glacier_vault_test.go [hunk 5]
```diff
@@ -239,47 +348,119 @@ func testAccCheckGlacierVaultDestroy(s *terraform.State) error {
...
+    ]
+}
+EOF
+}
+`, rName)
+}
+
+func testAccGlacierVaultPolicyConfigUpdated(rName string) string {
+	return fmt.Sprintf(`
+data "aws_region" "current" {}
```
Comment (bflad) [inline aws/resource_aws_glacier_vault_test.go:411]:
```suggestion
data "aws_partition" "current" {}

data "aws_region" "current" {}
```

File: aws/resource_aws_glacier_vault_test.go [hunk 6]
```diff
@@ -239,47 +348,119 @@ func testAccCheckGlacierVaultDestroy(s *terraform.State) error {
...
+             "AWS": ["*"]
+          },
+          "Effect":"Allow",
+          "Action": [
+             "glacier:UploadArchive",
+             "glacier:InitiateMultipartUpload",
+             "glacier:AbortMultipartUpload",
+             "glacier:CompleteMultipartUpload"
+          ],
+          "Resource": ["arn:aws:glacier:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:vaults/%[1]s"]
```
Comment (bflad) [inline aws/resource_aws_glacier_vault_test.go:434]:
```suggestion
          "Resource": ["arn:${data.aws_partition.current.partition}:glacier:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:vaults/%[1]s"]
```

**Review by bflad — COMMENTED** [review]
Appease newer static analysis

**Pull request merged**